### PR TITLE
fix(site): ensure Error Boundary catches render errors correctly

### PR DESCRIPTION
**Pull request by buenos-nachos**
No issue or design to link to. Squeezing this in between other stuff

## Changes made
- Replaced previous `ErrorBoundary` functionality with `GlobalErrorBoundary` component
- Wired up `GlobalErrorBoundary` to React Router rather than the top of the app

## Screenshots
Before
Screenshot 2024-12-26 at 5 49 54 PM

After
Screenshot 2024-12-26 at 5 50 25 PM

Screenshot 2024-12-26 at 5 50 36 PM

## Notes
- The visual design of the error boundary (and maybe some of the functionality) _should_ definitely change to be in line with our design kit, but I think that can wait. I did try to mirror our old and new design tokens as closely as possible (erring on the side of the old if there were conflicts), but to honest, I think the most important thing is that our error-handling functionality is now working correctly for the first time in ~2 years

### CODE REVIEW

File: site/src/components/Button/Button.tsx [hunk 1]
```diff
@@ -51,9 +51,12 @@ export const Button: FC<ButtonProps> = forwardRef<
 	const Comp = asChild ? Slot : "button";
 	return (
 		<Comp
-			className={cn(buttonVariants({ variant, size, className }))}
 			ref={ref}
 			{...props}
+			className={cn(buttonVariants({ variant, size, className }))}
+			type={
+				props.type === undefined && Comp === "button" ? "button" : props.type
+			}
```
Comment (Parkreiner) [inline site/src/components/Button/Button.tsx:59]: Small change to make sure that you have to opt into submit behavior for true buttons

File: site/src/components/Button/Button.tsx [hunk 2]
```diff
@@ -51,9 +51,12 @@ export const Button: FC<ButtonProps> = forwardRef<
 	const Comp = asChild ? Slot : "button";
 	return (
 		<Comp
-			className={cn(buttonVariants({ variant, size, className }))}
 			ref={ref}
 			{...props}
+			className={cn(buttonVariants({ variant, size }), className)}
```
Comment (Parkreiner) [inline site/src/components/Button/Button.tsx:56]: We had a typo where we were accidentally passing the consumer class name into CVA, instead of passing it directly to `cn`

File: site/src/components/ErrorBoundary/GlobalErrorBoundary.tsx [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+						</g>
+						<defs>
+							<clipPath id="clip0_103_80">
+								<rect width="68" height="49" />
+							</clipPath>
+						</defs>
+					</svg>
+
+					<div className="text-content-primary flex flex-col gap-1">
+						<h1 className="text-2xl font-normal m-0">{errorPageTitle}</h1>
```
Comment (Parkreiner) [inline site/src/components/ErrorBoundary/GlobalErrorBoundary.tsx:75]: @BrunoQuaresma I was wondering about this: I swear Tailwind's preflight styles get rid of all margins, period. So I wasn't sure why I as still getting them for headers and paragraphs

I saw a few other components using this, but I wasn't sure if we had to override Tailwind to get it to work with our current MUI setup

File: site/src/components/Button/Button.tsx (hunk 1 above)
Comment (BrunoQuaresma) [reply site/src/components/Button/Button.tsx:59]: I’m not sure I fully understand the value or use case for this 🤔. Could you share an example to help clarify? At first glance, it seems like the type prop should be the consumer’s responsibility rather than something built into the component.

File: site/src/components/ErrorBoundary/GlobalErrorBoundary.tsx (hunk 3 above)
Comment (BrunoQuaresma) [reply site/src/components/ErrorBoundary/GlobalErrorBoundary.tsx:75]: Currently, we’ve disabled Tailwind’s preflight option to simplify integration with MUI. Hopefully, we can enable it again soon.

**References:**
- [Slack conversation related to the decision](https://codercom.slack.com/archives/CJURPL8DN/p1729012756931799?thread_ts=1729001742.019449&cid=CJURPL8DN)
- [Tailwind’s config file where the preflight option is disabled](https://github.com/coder/coder/blob/main/site/tailwind.config.js#L4)

File: site/src/components/ErrorBoundary/GlobalErrorBoundary.tsx [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+						</g>
+						<defs>
+							<clipPath id="clip0_103_80">
+								<rect width="68" height="49" />
+							</clipPath>
+						</defs>
+					</svg>
```
Comment (BrunoQuaresma) [inline site/src/components/ErrorBoundary/GlobalErrorBoundary.tsx:72]: **Nit:** Since you’re already working on this part of the code, maybe we could replace this long SVG with the Coder logo component we use for the navbar. What do you think?

**Comment by BrunoQuaresma** [comment]
Awesome work, and thanks so much for getting this fixed 🙏. The only blocker for me is the decision in the `Button` component regarding passing a default value for `type`.

File: site/src/components/Button/Button.tsx (hunk 1 above)
Comment (Parkreiner) [reply site/src/components/Button/Button.tsx:59]: So, I feel that +90% of the time, when you have a `<button>`, you want it to be `type="button"` instead of the default `type="submit"`, and even experienced frontend engineers can forget to do that. That sometimes leads to forms accidentally getting submitted when you didn't mean for them to. I think it's a more reasonable default value that can't ever be made official because of HTML backwards-compatibility

And since the Radix button component is polymorphic, we don't have any linting rules to catch that you didn't set the type, but also because it's polymorphic, the final output might not be an actual button. If we use `asChild` to style a link like a button, we can't always assume that we can attach a `type` attribute. `<a>` doesn't support the prop, so while React would let us do it no problem, the HTML would be invalid

So basically, the check makes it so that buttons will default to `type="button"` if the output would be a button, and if the prop wasn't already set. Otherwise, yeah, it does fall to the user to wire things up correctly

File: site/src/components/ErrorBoundary/GlobalErrorBoundary.tsx (hunk 4 above)
Comment (Parkreiner) [reply site/src/components/ErrorBoundary/GlobalErrorBoundary.tsx:72]: No, not a nit at all. This component was originally code for the registry that I repurposed, and I forgot to see if we had an existing logo we could swap in

I'll get that fixed

File: site/src/components/Button/Button.tsx (hunk 1 above)
Comment (BrunoQuaresma) [reply site/src/components/Button/Button.tsx:59]: Do you think it would make sense to summarize this decision and add it as a comment? I’m concerned we might forget this decision in the future and accidentally remove it, as it’s not a very common or usual approach.

**Review by BrunoQuaresma — APPROVED** [review]

**Review by aqandrew — APPROVED** [review]
